covid-projections/covid-projections
# add error boundaries around location page blocks

[Error Boundaries](https://reactjs.org/docs/error-boundaries.html) catch JavaScript errors in children components and prevents the error from breaking the entire app. Currently, we have an error boundary in the Router, so if a feature anywhere in a page throws an error, we show "Something went wrong" on the entire page.

![location page showing "something went wrong"](https://user-images.githubusercontent.com/114084/110685512-37b95300-8193-11eb-84cb-d416479c94d5.png)

For example, this happened when we forgot to include vaccination data for the West Virginia ([see error in Sentry](https://sentry.io/organizations/covidactnow/issues/2227716089/events/bef2ca4c507f4a749d450597e2799401/)). This PR implements a component that includes the error boundary to wrap location page blocks. 

```tsx
<LocationPageBlock>
  <VulnerabilitiesBlock scores={ccviScores} region={region} />
</LocationPageBlock>
```

In this case, if the `VulnerabilitiesBlock` component crashes, it will show an error message and log the error to Sentry, but all the other blocks will render correctly.

![the error message shows only for the feature block](https://user-images.githubusercontent.com/114084/110685853-8ff05500-8193-11eb-906e-c269e9f6d8e4.png)

## Testing

This will only be shown in production builds, in development React renders a different component with the stack trace. To test:

- Simulate a feature crashing, add the following line to any feature
  ```tsx
  new URL('~~~');
  ```
- Build and serve the app in production mode by running `yarn serve`. 
- Open a location page. You should see the error message on the feature only, instead of the full-page "something went wrong" messsage.
